chapter8.adoc

```diff
@@ -1,27 +1,66 @@
 [[chapter8]]
 [[Smsdedbg]]
-== `Smsdedbg`: Supervisor Domain External Debug
+== Supervisor Domain External Trace and Debug
 
-The <<MSDCFG>> CSR contains the `sdedbgalw` bit that controls whether the
-current scheduled SD is allowed to be external-debugged cite:[ExtDbg]. This bit
-is context switched (along with rest of the `msdcfg`) per SD.
+The <<MSDCFG>> CSR contains the `sdedbgalw` and `sdetrcalw` bits that control
+whether the current scheduled SD is allowed to be external-debugged or to be
+external-traced respectively.
 
-=== `Smsdedbg` interaction with external debug security controls (Informative)
+=== `Smsdedbg`: External Debug allowed control for Supervisor Domain
+
+When M-mode external debug is enabled, all supervisor domains may also be
+debugged by an external debugger irrespective of the configuration held in
+`msdcfg.SDEDBGALW`.
+
+When M-mode external debug is disabled, whether execution at privilege modes
+less than `M-mode` may be debugged by an external debugger depends on the
+configuration held in `msdcfg.SDEDBGALW`.
+
+When `msdcfg.SDEDBGALW` = 0, external debug is disallowed for the SD. Abstract
+commands and halt request from the debug module are suppressed and stay pending
+while the SD is active.
+
+When `msdcfg.SDEDBGALW` = 1 then external debug of privilege modes less than
+`M-mode` is allowed for the SD, and:
+
+* A halt request may transition the hart to Debug Mode.
+* Abstract commands and program buffer execution can access state of privilege
+modes less than `M-mode`.
+* Read and Write of `Sdtrig` and `Sdext` CSRs are allowed.
```

Comment:
so the statement is qualified by the title of the header of the bullet points - specifically - "When `msdcfg.SDEDBGALW` = 1 then external debug of privilege modes less than `M-mode` is allowed for the SD on a per-hart basis, and:" -> Read and Write of `Sdtrig` and `Sdext` CSRs are allowed . 
Once the SD opts in to external debug, all rules specified in debug spec for CSR accesses by the SD or by M on behalf of the SD apply so we don't have to restate those IMO.

Reply:
It introduce security holes. If you look at sdtrig CSR tdata1. It could set trigger for machine mode (e.g. set trigger to raise exception in M-mode when it is not yet context switched). In your statement, it just expose the privilege escalation attack vector. 
BTW, what do you mean by "M on behalf of the SD"?

Reply:
"M on behalf of the SD" -> means RDSM enforcing any rules since it is in the TCB of the SDs.
an e.g. for this is -  we just added a note that for any triggers a SD sets up, RDSM must ensure the m bit is cleared so that prevents the scenario you mentioned.

can you clarify what security holes you see?

Reply:
Sure, of course. What I mentioned above is exactly an example of hole and it cannot be prevented by RDSM, because the external debugger set it without involvement of RDSM.

Reply:
M-mode needs to protect itself in this case. which is we have the new proposed behavior change (that I hope you will pull into the ext secure debug spec), specifically:

When `medbgen` is 0 and privilege is `M-mode`:
* M-mode can read and write triggers, *including triggers with `dmode` = 1 without restrictions*.
* Abstract commands and requests from an external debugger, that transition the hart into a halted state, stay pending while privilege is `M-mode` cite:[ExtDbg].
* All trigger-matching is prevented from matching or firing, similar to how `MIE` or `MTE` would behave.

Reply:
`MIE` and `MTE` solution is to tackle re-entrance problem (with action=0), while the issue I mentioned in my previous post is about the trigger traps into M-mode. Re-entrance issue only happens after it. 
More importantly, the statement "similar to how MIE or MTE would behave." is misleading. I thought it means:
- similar to action=0, the triggers with action=1 are prevented from firing/matching to enter Debug Mode

But based on your comment, seems it is supposed to convey different idea:

- triggers with action=1 are prevented from enter debug mode in M-mode
- triggers with action=0 are prevented from raising exception in M-mode

The second bullet is brand new idea and never bought up. I think we need involve more people discuss about the scope it affected. 

Above all, "similar to how MIE or MTE would behave." has to be taken out. It gives misleading idea.

Reply:
"All trigger-matching" - could implies address/data/exception/interrupt matching (which is arguably intuitive for readers). It is not a good point to keep ambiguous statement based on personal understanding. 

It is a brand new in the scope of external debug security.
If you quote the whole context, it could be easily identified that the Debug spec introduce them to solve re-entrance issue, which prevents exception from same privilege (e.g. M-mode->M-mode).
But in scope of external debug, the implication is to prevent the exception from lower privilege (e.g. S-mode -> M-mode). Both the hardware implementation and software use case are different.

Reply:
If I didn't make the point clear enough, we could probably discuss it further in tomorrow's external debug TG meeting? 

[https://zoom.us/j/99483621157?pwd=cXYvT0I4QXBMdEZlVkhjbnRtUEdrUT09](https://www.google.com/url?q=https://zoom.us/j/99483621157?pwd%3DcXYvT0I4QXBMdEZlVkhjbnRtUEdrUT09&sa=D&source=calendar&usd=2&usg=AOvVaw3-CCTBCDyWhEIIJEWGPueR)

Reply:
I suggest using the precise words from section 5.4 and section 5.7.6, repeated above, rather than loose re-interpretations like  "which prevents exception from same privilege".

Reply:
Before driving this topic off-track, I would strongly suggest let's discuss it on TG meeting . Honestly, I don't think we are talking about the same thing.

Reply:
I added normative references to section 5.4 and 5.7.6 of Debug spec.